BondageClub/Scripts/ImageCache.js

```diff
@@ -0,0 +1,217 @@
+"use strict";
+
+/**
+ * What state is the image in.
+ * @enum string
+ */
+const CachedImageState = {
+	UNCACHED: 'uncached',
+	LOADING: 'loading',
+	LOADED: 'loaded',
+	ERRORED: 'errored',
+	FAILED: 'failed',
+};
+
+/**
+ * Cached image data
+ */
+class CachedImage {
+	/**
+	 * Create a new cached image.
+	 *
+	 * @param {ImageCache} cache - The cache holding the image
+	 * @param {string} url - The URL for the image to cache
+	 * @param {function(CachedImage):void} [loadCallback] - A function called when the image loads
+	 */
+	constructor(cache, url, loadCallback = null) {
+		/**
+		 * Only used to inform the cache of loading progress
+		 * @type ImageCacheClass
+		 * @private
+		 */
+		this.cache = cache;
+		/** @type string */
+		this.url = url;
+		/** @type CachedImageState */
+		this.state = CachedImageState.UNCACHED;
+		/** @type number */
+		this.errorCount = 0;
+		/** @type {function(CachedImage):void} */
+		this.loadCallback = loadCallback;
+		/** @type HTMLImageElement */
+		this.element = null;
+		/** @type {{ width: number; height: number; texture: WebGLTexture; }} */
+		this.textureInfo = null;
+	}
+
+	/**
+	 * Load the cached image if it's not already available
+	 */
+	load() {
+		if ((this.state == CachedImageState.LOADING) || (this.state == CachedImageState.LOADED))
+			return;
+
+		// Get a new image element to perform the loading for us
+		this.element = new Image();
+
+		// Set up listeners to keep track of the loading process
+		let image = this;
+		this.element.addEventListener("load", function() {
+			image.errorCount = 0;
+			image.state = CachedImageState.LOADED;
+			image.cache._imageStateDidChange(image);
+		});
```

Comment:
I'd drop the `this` aliasing completely and just use arrows in the callback. Similarly for the listener below.
```suggestion
		this.element.addEventListener("load", () => {
			this.errorCount = 0;
			this.state = CachedImageState.LOADED;
			this.cache._imageStateDidChange(image);
		});
```

Reply:
I think I checked, and `this` wasn't the correct object (`window`, IIRC), hence the "weak" reference. I'll check again and fix if needed.

Reply:
`this` won't be the correct object if you pass the callback in as a `function` (i.e. `function() {}`, since functions have their own scope. But if you pass the callback in as an arrow function, as arrows inherit the ambient scope. So:
```js
const outsideThis = this;

// This function won't have access to this
const func = function() {
    this === outsideThis; // false
}

// But this function will
const arrow = () => {
    this === outsideThis; // true
}
```

Reply:
Fixed, thanks for the pointer ❤️